### PR TITLE
Compare nodes in addition to symRef in SequentialStoreSimplification

### PR DESCRIPTION
**Pull request by rmnattas**
SequentialStoreSimplification compares baseVar (first child to `aladd`) symRef to confirm stores being to the same array for simplification. For object fields the symRef matches even if arrays are different.

To fix the issue a node check is added to make sure that its storing to the same array even if symRefs match.

```
n64n      bstorei  <array-shadow>[#222  Shadow] [flags 0x80000601 0x0 ]                       [0x7284a30603c0] bci=[-1,46,2542] rc=0 vc=730 vn=- li=- udi=- nc=2
n63n        aladd (X>=0 internalPtr )                                                         [0x7284a3060370] bci=[-1,46,2542] rc=1 vc=730 vn=- li=- udi=- nc=2 flg=0x8100
n8n           aloadi  TestSeqStore.a [B[#419  Shadow +8] [flags 0x607 0x0 ]                     [0x736004404ac0] bci=[-1,1,5] rc=6 vc=2185 vn=- li=- udi=- nc=1
n7n             aload  <parm 0 LTestSeqStore;>[#417  Parm] [flags 0x40000107 0x0 ] (X>=0 )      [0x736004404a70] bci=[-1,0,5] rc=2 vc=2185 vn=- li=- udi=13 nc=0 flg=0x100
n62n          lconst 9 (highWordZero X!=0 X>=0 )                                              [0x7284a3060320] bci=[-1,46,2542] rc=1 vc=730 vn=- li=- udi=- nc=0 flg=0x4104
n56n        bconst  93 (X!=0 X>=0 )                                                           [0x7284a3060140] bci=[-1,46,2542] rc=1 vc=730 vn=- li=- udi=- nc=0 flg=0x104
n76n      bstorei  <array-shadow>[#222  Shadow] [flags 0x80000601 0x0 ]                       [0x7284a3060780] bci=[-1,51,2543] rc=0 vc=730 vn=- li=- udi=- nc=2
n75n        aladd (X>=0 internalPtr )                                                         [0x7284a3060730] bci=[-1,51,2543] rc=1 vc=730 vn=- li=- udi=- nc=2 flg=0x8100
n19n          aloadi  TestSeqStore.a [B[#419  Shadow +8] [flags 0x607 0x0 ]                     [0x736004404e30] bci=[-1,10,5] rc=4 vc=2185 vn=- li=- udi=- nc=1
n18n            aload  <parm 1 LTestSeqStore;>[#418  Parm] [flags 0x40000107 0x0 ] (X>=0 )      [0x736004404de0] bci=[-1,9,5] rc=1 vc=2185 vn=- li=- udi=14 nc=0 flg=0x100
n74n          lconst 10 (highWordZero X!=0 X>=0 )                                             [0x7284a30606e0] bci=[-1,51,2543] rc=2 vc=730 vn=- li=- udi=- nc=0 flg=0x4104
n68n        bconst  13 (X!=0 X>=0 )                                                           [0x7284a3060500] bci=[-1,51,2543] rc=1 vc=730 vn=- li=- udi=- nc=0 flg=0x104
```

### CODE REVIEW

**Comment by rmnattas** [comment]
fyi @zl-wang @hzongaro
Testing PR

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1659]: In general, it is good practice to call `hasSymbolReference` first before calling `getSymbolReference` 

That pattern is seen in many other places.

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1675]: Same comment as before

**Comment by vijaysun-omr** [comment]
I will wait for @hzongaro review and the WIP to be taken off before I consider merging.

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (rmnattas) [reply runtime/compiler/optimizer/SequentialStoreSimplifier.cpp]: Fixed. Made an if-block to remove redundant checks.

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (hzongaro) [inline runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1689]: I think the comparison of `baseRef` and `activeBaseRef` is now redundant.  `baseNode == activeBaseNode` implies `baseRef == activeBaseRef`, while `baseRef != activeBaseRef` implies `baseNode != activeBaseNode`.

Comment (rmnattas) [reply runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1689]: Make sense, then no reason to get symRefs in the first place.
The other usage of symRef in the code is to compare it to in the value tree.
https://github.com/eclipse-openj9/openj9/blob/73af0c72ea6d61f98bc7b07ff2d6c02128a71d61/runtime/compiler/optimizer/SequentialStoreSimplifier.cpp#L1709-L1714

Which would need to be changed too to verify using nodes instead of symRefs unless it will be too conservative...
A similar recursive tree lookup for nodes would be needed to use instead (if exists?).

**Comment by rmnattas** [comment]
Sequential Load Simplification is a part of `SequentialStoreSimplification` that @IBMJimmyk worked on.
He mentioned that it already does node comparison to verify base variable but for OffHeap it wouldn't allow tree shapes accessing first element given that it doesn't have an `aladd`.

I added another commit to this PR that solves that.

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (IBMJimmyk) [inline runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1071]: For the sequential load simplification part of the opt, this check guards the rest of the opt. From here on, it assumes the `aladd` node exists and won't check again. If you bypass it like this, other things can break.

These two definitely won't work anymore:
getALoadReferenceForSeqLoad:
https://github.com/eclipse-openj9/openj9/blob/73af0c72ea6d61f98bc7b07ff2d6c02128a71d61/runtime/compiler/optimizer/SequentialStoreSimplifier.cpp#L1362-L1402
getBaseOffsetForSeqLoad:
https://github.com/eclipse-openj9/openj9/blob/73af0c72ea6d61f98bc7b07ff2d6c02128a71d61/runtime/compiler/optimizer/SequentialStoreSimplifier.cpp#L1231-L1289

There may be others.

**Comment by vijaysun-omr** [comment]
May I request some printfs in the code to tell you how many times the sequential store elimination did transformations in a couple of the benchmarks that would run the opt with and without your changes ? If the opt only runs at hot and above, then you probably won't be able to run Liberty scenarios to gauge the effectiveness, and would need some other apps like SPECjbb2015 or ILOG or Spark. 

I understand the optimization was doing a potentially incorrect transformation, but there may have been a case where it did not have commoned nodes and was still transforming correctly, but now no longer does so. I don't want to delay your changes too much, so you can fix the functional issue first, and then follow up with performance evaluation in the near term and that would be okay with me.

**Comment by rmnattas** [comment]
I'll lookup on the load simplification changes necessary and may decide to put do it separately after instead to get this fix going. Load simplification will be functionally correct but not being able to transform the specific case of first element accesses.

I'll add the printfs and measure it after completing the fix.

**Comment by rmnattas** [comment]
PR ready for merge. Testing performance I don't see difference in running benchmarks.

**Comment by vijaysun-omr** [comment]
Jenkins test sanity.functional,sanity.openjdk jdk21

**Comment by hzongaro** [comment]
Jenkins test sanity.functional,sanity.openjdk xlinux jdk21

**Comment by vijaysun-omr** [comment]
Testing has passed, @hzongaro are you thinking of approving this ? I ask, to check before I consider a merge

**Review by hzongaro — APPROVED** [review]
I think this looks good.  Thanks!

**Comment by hzongaro** [comment]
> are you thinking of approving this?

Sorry for the delay - I had just finished reviewing.

File: runtime/compiler/optimizer/SequentialStoreSimplifier.cpp
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1711]: I wonder if we should leave this spot in the code unchanged. In the other places changed in this PR, node equality makes things more conservative, whereas I think in this place, it makes things less conservative than was the case before.

Comment (rmnattas) [reply runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1711]: I agree that it makes it less conservative, but thinking that it will be too conservative with the code not changed.
For field stores, any access of that field in the value tree (no matter the base array) would invalidate the transformation as symRefs will match.

Comment (rmnattas) [reply runtime/compiler/optimizer/SequentialStoreSimplifier.cpp:1711]: I think I see the issue, with node comparison if two access the same field and baseNode is different but might have the same base array value, this would allow it when it shouldn't.

**Pull request merged**
